File: NetFramework/TouchToolkit/VB/TouchToolkitDemo/GcZoomPages/Help UserControl/MyTrackBar.vb
Imports System.Collections.Generic
Imports System.ComponentModel
Imports System.Drawing
Imports System.Data
Imports System.Linq
Imports System.Text
Imports System.Windows.Forms

Namespace C1ZoomPages.Help_Forms
	Public Partial Class MyTrackBar
		Inherits UserControl
		Public Sub New()
			InitializeComponent()

			MaxValue = 4
			MinValue = 1

			AddHandler button2.LocationChanged, AddressOf button2_LocationChanged
			AddHandler button2.MouseMove, AddressOf button2_MouseMove
			AddHandler button2.MouseDown, AddressOf button2_MouseDown
			AddHandler button2.MouseUp, AddressOf button2_MouseUp
			AddHandler button2.MouseCaptureChanged, AddressOf button2_MouseCaptureChanged
		End Sub

		Private downLocation As System.Nullable(Of Integer) = Nothing
		Private startY As Integer = 0

		Private Sub button2_MouseCaptureChanged(sender As Object, e As EventArgs)
			downLocation = Nothing
		End Sub

		Private Sub button2_MouseUp(sender As Object, e As MouseEventArgs)
			downLocation = Nothing
		End Sub

		Private Sub button2_MouseDown(sender As Object, e As MouseEventArgs)
			downLocation = button2.PointToScreen(e.Location).Y
			startY = button2.Location.Y
		End Sub

		Private Sub button2_MouseMove(sender As Object, e As MouseEventArgs)
			Dim moveLocation = button2.PointToScreen(e.Location).Y

			If downLocation IsNot Nothing Then
				Dim y As Integer = startY + (moveLocation - downLocation.Value)
				If y < 0 Then
					y = 0
				End If
				If y > button1.Height - button2.Height Then
					y = button1.Height - button2.Height
				End If
				button2.Location = New Point(button2.Location.X, y)
			End If
		End Sub

		Private Sub button2_LocationChanged(sender As Object, e As EventArgs)
			OnValueChanged(e)
		End Sub

		<Bindable(True)> _
		Public Property Value() As Single
			Get
				Return CSng(button2.Location.Y) / (button1.Height - button2.Height) * (MaxValue - MinValue) + MinValue
			End Get
			Set
				button2.Location = New Point(button2.Location.X, CInt((button1.Height - button2.Height) * (value - MinValue) / (MaxValue - MinValue)))
			End Set
		End Property

		Public Property MaxValue() As Single
			Get
				Return m_MaxValue
			End Get
			Set
				m_MaxValue = Value
			End Set
		End Property
		Private m_MaxValue As Single
		Public Property MinValue() As Single
			Get
				Return m_MinValue
			End Get
			Set
				m_MinValue = Value
			End Set
		End Property
		Private m_MinValue As Single

		Public Event ValueChanged As EventHandler(Of EventArgs)

		Protected Overridable Sub OnValueChanged(e As EventArgs)
			RaiseEvent ValueChanged(Me, e)
		End Sub
	End Class
End Namespace

File: NetFramework/TouchToolkit/CS/TouchToolkitDemo/GcZoomPages/Help UserControl/MyTrackBar.cs
using System;
using System.Collections.Generic;
using System.ComponentModel;
using System.Drawing;
using System.Data;
using System.Linq;
using System.Text;
using System.Windows.Forms;

namespace ClickOnceDemo.C1ZoomPages.Help_Forms
{
    public partial class MyTrackBar : UserControl
    {
        public MyTrackBar()
        {
            InitializeComponent();

            MaxValue = 4;
            MinValue = 1;

            button2.LocationChanged += button2_LocationChanged;
            button2.MouseMove += button2_MouseMove;
            button2.MouseDown += button2_MouseDown;
            button2.MouseUp += button2_MouseUp;
            button2.MouseCaptureChanged += button2_MouseCaptureChanged;
        }

        int? downLocation = null;
        int startY = 0;

        void button2_MouseCaptureChanged(object sender, EventArgs e)
        {
            downLocation = null;
        }

        void button2_MouseUp(object sender, MouseEventArgs e)
        {
            downLocation = null;
        }

        void button2_MouseDown(object sender, MouseEventArgs e)
        {
            downLocation = button2.PointToScreen(e.Location).Y;
            startY = button2.Location.Y;
        }

        void button2_MouseMove(object sender, MouseEventArgs e)
        {
            var moveLocation = button2.PointToScreen(e.Location).Y;

            if (downLocation != null)
            {
                int y = startY + (moveLocation - downLocation.Value);
                if (y < 0)
                {
                    y = 0;
                }
                if (y > button1.Height - button2.Height)
                {
                    y = button1.Height - button2.Height;
                }
                button2.Location = new Point(button2.Location.X, y);
            }
        }

        void button2_LocationChanged(object sender, EventArgs e)
        {
            OnValueChanged(e);
        }

        [Bindable(true)]
        public float Value
        {
            get
            {
                return (float)(button2.Location.Y) / (button1.Height - button2.Height) * (MaxValue - MinValue) + MinValue;
            }
            set
            {
                button2.Location = new Point(button2.Location.X, (int)((button1.Height - button2.Height) * (value - MinValue) / (MaxValue - MinValue)));
            }
        }

        public float MaxValue { get; set; }
        public float MinValue { get; set; }

        public event EventHandler<EventArgs> ValueChanged;

        protected virtual void OnValueChanged(EventArgs e)
        {
            if (this.ValueChanged != null)
            {
                this.ValueChanged(this, e);
            }
        }
    }
}
